### PR TITLE
binary-dist configure/Makefile handling

### PR DESCRIPTION
**Pull request by alpmestan**
Note: this requires a patch of mine on top of the llvmng branch of ghc: https://github.com/alpmestan/ghc/commit/363690a124a67ba5a6073d09d0b5f842d8fc2f73

### CODE REVIEW

File: src/Rules/Register.hs [hunk 1]
```diff
@@ -83,6 +83,7 @@ buildConf _ context@Context {..} _conf = do
            , bldPath -/- "ghcautoconf.h"
            , bldPath -/- "ghcplatform.h"
            , bldPath -/- "ghcversion.h"
+           , bldPath -/- "ffi.h"
```
Comment (angerman) [inline src/Rules/Register.hs:86]: Is this still needed? I might be misremember something here.

File: src/Rules.hs [hunk 2]
```diff
@@ -46,10 +46,30 @@ topLevelTargets = do
...
+      -- copy config.sub, config.guess, install-sh, Makefile files, etc
+      -- from the source of the tree to the bindist dir
+      copyFile (cwd -/- "distrib" -/- "configure") (baseDir -/- "configure")
+      copyFile (cwd -/- "distrib" -/- "Makefile") (baseDir -/- "Makefile")
+      copyFile (cwd -/- "install-sh") (baseDir -/- "install-sh")
+      copyFile (cwd -/- "config.sub") (baseDir -/- "config.sub")
+      copyFile (cwd -/- "config.guess") (baseDir -/- "config.guess")
+      copyFile (cwd -/- "settings.in") (baseDir -/- "settings.in")
+      copyFile (cwd -/- "mk" -/- "config.mk.in") (baseDir -/- "mk" -/- "config.mk.in")
+      copyFile (cwd -/- "mk" -/- "install.mk.in") (baseDir -/- "mk" -/- "install.mk.in")
```
Comment (angerman) [inline src/Rules.hs:64]: The rest of the build system uses `<~` for those copy rules (in a `where` clause though). Maybe you want to use that here as well?

File: shell.nix [hunk 3]
```diff
@@ -57,5 +57,7 @@ in
                     nixpkgs.ncurses
                     nixpkgs.m4
                     nixpkgs.gmp
-                    nixpkgs.file ];
+                    nixpkgs.file
+		    nixpkgs.llvm_5
```
Comment (angerman) [inline shell.nix]: indent!

**Review by angerman — APPROVED** [review]
Feel free to merge when fixed up.

File: src/Rules/Register.hs (hunk 1 above)
Comment (alpmestan) [reply src/Rules/Register.hs:86]: Oh I didn't notice that I kept this around. Well, let's try removing it...

File: shell.nix (hunk 3 above)
Comment (alpmestan) [reply shell.nix]: oopsies.

File: src/Rules/Register.hs (hunk 1 above)
Comment (alpmestan) [reply src/Rules/Register.hs:86]: The build breaks on my machine when I remove this line.

File: src/Rules.hs (hunk 2 above)
Comment (alpmestan) [reply src/Rules.hs:64]: Apparently that's only used in `Rules.Generate` and only defined locally indeed. I'd suggest maybe keeping the explicit `copyFile`, but doing a `forM` over the list of files to copy over in `baseDir`?

File: shell.nix (hunk 3 above)
Comment (alpmestan) [reply shell.nix]: fixed

**Comment by alpmestan** [comment]
@angerman Could you possibly try this branch, but also adding the following `Makefile` under `distrib/` before building/running hadrian etc?

``` Makefile
MAKEFLAGS += --no-builtin-rules
.SUFFIXES:

ProjectVersion:=$(shell bin/ghc --numeric-version)

include mk/install.mk

define GHC_WRAPPER
#!/bin/sh
exec "$(libdir)/bin/ghc" -B"$(libdir)" $${1+"$$@"}
endef

export GHC_WRAPPER

.PHONY: default
default:
	@echo 'Run "make install" to install'
	@false

.PHONY: install
install:
	@cp settings lib/
	@echo $(prefix) $(bindir) $(libdir)
	@mkdir -p $(prefix) $(libdir) $(bindir)
	@cp -R lib/* $(libdir)/
	# cp the rest to $(prefix) ? or maybe handle $(datadir) etc too?
	@if [ "$(bindir)" = "$(prefix)/bin" ] || [ "$(libdir)" != "$(prefix)/lib" ]; then \
	   echo "custom bindir or libdir"; \
	   mkdir -p $(libdir)/bin; \
	   cp bin/* $(libdir)/bin/; \
	   echo "$$GHC_WRAPPER" > $(bindir)/ghc; \
	fi
	@echo "ghc available at $(bindir)/ghc"
	@echo done
```

So, you build the `binary-dist` target with hadrian, take the archive and `./configure`/`make install` as you'd normally do with a binary dist.

**Pull request merged**